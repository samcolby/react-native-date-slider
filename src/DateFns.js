import moment from "moment";

function initLocale(name, config) {
  try {
    moment.locale(name, config);
  } catch (error) {
    throw new Error(
      "Locale prop is not in the correct format. \n Locale has to be in form of object, with keys of name and config. " +
        error.message
    );
  }
}

function initMoment(date, locale) {
  if (locale && locale.name) {
    return moment(date).startOf("day").locale(locale.name);
  } else {
    return moment(date).startOf("day");
  }
}

export default { initLocale, initMoment };
