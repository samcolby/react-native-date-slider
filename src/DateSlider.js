import React from "react";
import { Dimensions } from "react-native";

import _ from "lodash";

import DateFns from "./DateFns";

import DateSliderComponent from "./DateSliderComponent";

const END_THRESHOLD = 4;
const EXTEND_WEEKS_BY = 8;

class DateSlider extends React.Component {
  static propTypes = {
    isoStartDay: React.PropTypes.number,
    locale: React.PropTypes.object,
    onDateSelected: React.PropTypes.func,
    onWeekChanged: React.PropTypes.func,
    selectedDate: React.PropTypes.object
  };

  static defaultProps = {
    isoStartDay: 1
  };

  flatlist = undefined;
  selectedIndex = 0;
  viewableIndex = -1;
  isExpanding = false;

  constructor(props) {
    super(props);

    if (props.locale && props.locale.name && props.locale.config) {
      DateFns.initLocale(props.locale.name, props.locale.config);
    }

    let selectedDate;
    if (this.props.selectedDate) {
      selectedDate = DateFns.initMoment(this.props.selectedDate, props.locale);
    } else {
      selectedDate = DateFns.initMoment(new Date(), props.locale);
    }

    // The very first date in the FlatList
    const startingDate = DateFns.getStartOfWeek(
      selectedDate,
      props.isoStartDay
    ).subtract(EXTEND_WEEKS_BY, "weeks");

    this.state = {
      viewWidth: Dimensions.get("window").width,
      arrDates: this._generateDates(
        startingDate,
        EXTEND_WEEKS_BY * 2 + 1,
        selectedDate,
        props.isoStartDay
      ),
      selectedDate: selectedDate
    };
  }

  componentWillReceiveProps(nextProps) {
    if (
      nextProps.selectedDate &&
      !this.state.selectedDate.isSame(nextProps.selectedDate, "day")
    ) {
      this._setSelectedDate(
        DateFns.initMoment(nextProps.selectedDate, nextProps.locale)
      );
      this._scrollTo(this.selectedIndex);
    }
  }

  render() {
    return (
      <DateSliderComponent
        arrDates={this.state.arrDates}
        captureFlatListRef={this._captureFlatListRef}
        viewWidth={this.state.viewWidth}
        getItemLayout={this._getItemLayout}
        onDateSelected={this._onDateSelected}
        onEndReached={this._onEndReached}
        onEndReachedThreshold={END_THRESHOLD}
        onLayout={this._onLayout}
        onViewableItemsChanged={this._onViewableItemsChanged}
      />
    );
  }

  /**
   * Pure function to generate an array of objects that can be used by the FlatList.
   *
   * The FlatList renders a 'week' as a new item. This function
   * creates a number of entries equal to the weeks parameter
   * and starting from the fromDate paramater. The fromDate
   * needs to be set to the correct weekday used to start each WeekView
   * by the callee.
   *
   * @param {moment} fromDate
   *  The starting date for the array of weeks to be generated
   * @param {number} weeks
   *  The number of weeks to be generated
   * @param {moment} selectedDate
   *  The date that is current selected
   * @param {number} isoStartDay
   *  The ISOWeekday number that starts the week being displayed (1 - 7)
   * @returns {Array<Object>}
   *
   * @memberof DateSlider
   */
  _generateDates(fromDate, weeks, selectedDate, isoStartDay) {
    const startOfSelectedWeek = DateFns.getStartOfWeek(
      selectedDate,
      isoStartDay
    );

    return _.map(_.range(weeks), i => {
      const weekStartDate = fromDate.clone().add(i, "weeks");
      if (startOfSelectedWeek.isSame(weekStartDate, "day")) {
        this.selectedIndex = i;
      }
      return {
        key: weekStartDate.toISOString(),
        date: weekStartDate,
        arrSelected: this._generateArrSelected(
          weekStartDate,
          startOfSelectedWeek,
          selectedDate,
          isoStartDay
        )
      };
    });
  }

  /**
   * A pure function to generate an array of boolean values
   * that specify if one of the days of the week is selected.
   *
   * This array is to be used as a prop on the WeekView component.
   *
   * @param {moment} weekStartDate
   *  The start date of the week this is to be generated
   * @param {moment} startOfSelectedWeek
   *  The start of the week that contains the selectedDate
   * @param {moment} selectedDate
   *  The date that is selected
   * @param {number} isoStartDay
   *  The ISOWeekday number that starts the week being displayed (1 - 7)
   *
   * @returns {Array<boolean>}
   *
   * @memberof DateSlider
   */
  _generateArrSelected(
    weekStartDate,
    startOfSelectedWeek,
    selectedDate,
    isoStartDay
  ) {
    let result = [false, false, false, false, false, false, false];
    if (weekStartDate && weekStartDate.isSame(startOfSelectedWeek, "day")) {
      result[DateFns.getSelectedDayOfWeek(selectedDate, isoStartDay)] = true;
    }
    return result;
  }

  /**
   * Captures the FlatList instance ref, so that
   * the scrollTo method can be called on it.
   * This is populated from the DateSliderComponent.
   *
   * @param {FlatList} flatlist
   *   The FlatList instance
   *
   * @memberof DateSlider
   */
  _captureFlatListRef = flatlist => (this.flatlist = flatlist);

  /**
   * ReactNative method called by the FlatList
   * This needs to be here as it uses viewWidth held in state
   *
   * @param {Object} data
   * @param {number} index
   *
   * @memberof DateSlider
   */
  _getItemLayout = (data, index) => {
    return {
      length: this.state.viewWidth,
      offset: this.state.viewWidth * index,
      index
    };
  };

  /**
   * ReactNative method called by the view containing the FlatList
   * This needs to be here as handles setting the viewWidth
   * based on what the containing view reports.
   *
   * @param {Object} event
   *  The event generated from the RN onLayout call
   *
   * @memberof DateSlider
   */
  _onLayout = event => {
    const nativeEvent = event.nativeEvent;
    this.setState({ viewWidth: nativeEvent.layout.width });
    if (this.viewableIndex === -1) {
      this._scrollTo(this.selectedIndex);
    } else {
      this._scrollTo(this.viewableIndex);
    }
  };

  /**
   * Scrolls the flatlist to the specified index.
   * This wraps the call in a setTimeout(_,0) Promise
   * to stop issues when the index to be scrolled to has not yet been
   * rendered.
   *
   * @param {number} index
   *  The index to scroll the flatlist to
   *
   * @memberof DateSlider
   */
  _scrollTo = index => {
    const scrollToPromise = new Promise(resolve => setTimeout(resolve, 0));
    scrollToPromise.then(() => {
      this.flatlist.scrollToIndex({
        viewPosition: 0,
        index: index,
        animated: false
      });
    });
  };

  _onEndReached = () => {
    this.isExpanding = true;
    const lastDate = this.state.arrDates[this.state.arrDates.length - 1].date;
    const appendWeeksFrom = lastDate.clone().add(1, "weeks");
    const arrNewDates = this._generateDates(
      appendWeeksFrom,
      EXTEND_WEEKS_BY,
      this.state.selectedDate,
      this.props.isoStartDay
    );
    this.setState({ arrDates: [...this.state.arrDates, ...arrNewDates] });
    this.isExpanding = false;
  };

  _onStartReached = () => {
    const firstDate = this.state.arrDates[0].date;
    const prependWeeksFrom = firstDate
      .clone()
      .subtract(EXTEND_WEEKS_BY, "weeks");
    const arrNewDates = this._generateDates(
      prependWeeksFrom,
      EXTEND_WEEKS_BY,
      this.state.selectedDate,
      this.props.isoStartDay
    );
    this.setState({ arrDates: [...arrNewDates, ...this.state.arrDates] });
    this.selectedIndex = this.selectedIndex + EXTEND_WEEKS_BY;
    this._scrollTo(END_THRESHOLD - 1 + EXTEND_WEEKS_BY);
    this.isExpanding = false;
  };

  _onViewableItemsChanged = ({ viewableItems, changed }) => {
    if (viewableItems.length === 1) {
      const viewableIndex = viewableItems[0].index;
      const changedIndex = changed[0].index;
      this.viewableIndex = viewableIndex;

      if (changedIndex === viewableIndex - 1) {
        // Week forward
        const date = this.state.arrDates[viewableIndex].date;
        this.props.onWeekChanged(date);
      }

      if (changedIndex === viewableIndex + 1) {
        // week back
        const date = this.state.arrDates[viewableIndex].date;
        this.props.onWeekChanged(date);
      }

      if (changedIndex <= END_THRESHOLD && viewableIndex === changedIndex - 1) {
        if (!this.isExpanding) {
          this.isExpanding = true;
          this._onStartReached();
        }
      }
    }
  };

  _onDateSelected = date => {
    this._setSelectedDate(date);
    this.props.onDateSelected(date);
  };

  _setSelectedDate = date => {
    const startOfSelectedWeek = DateFns.getStartOfWeek(
      date,
      this.props.isoStartDay
    );

    let arrDates = [...this.state.arrDates];

    const prevSelectedIndex = this.selectedIndex;

    const selectedIndex = _.findIndex(arrDates, [
      "key",
      startOfSelectedWeek.toISOString()
    ]);

    arrDates[prevSelectedIndex].arrSelected = this._generateArrSelected();
    arrDates[selectedIndex].arrSelected = this._generateArrSelected(
      arrDates[selectedIndex].date,
      startOfSelectedWeek,
      date,
      this.props.isoStartDay
    );

    this.selectedIndex = selectedIndex;

    this.setState({
      arrDates: arrDates,
      selectedDate: date
    });
  };
}

export default DateSlider;
