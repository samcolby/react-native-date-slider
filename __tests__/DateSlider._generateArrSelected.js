import moment from "moment";

import DateSlider from "../src/DateSlider";
import DateFns from "../src/DateFns";

const testData = [
  {
    weekStartDate: "2017-05-10",
    selectedDate: "2017-05-21",
    isoStartDay: 1
  },
  {
    weekStartDate: "2017-05-10",
    selectedDate: "2017-05-10",
    isoStartDay: 1
  },
  {
    weekStartDate: "2018-04-13",
    selectedDate: "2018-04-14",
    isoStartDay: 5
  },
  {
    weekStartDate: "1999-02-28",
    selectedDate: "1999-03-12",
    isoStartDay: 3
  },
  {
    weekStartDate: "2010-11-06",
    selectedDate: "2010-10-01",
    isoStartDay: 6
  },
  {
    weekStartDate: "2025-03-01",
    selectedDate: "2010-10-01",
    isoStartDay: 2
  },
  {
    weekStartDate: "2025-03-01",
    selectedDate: "2025-03-12",
    isoStartDay: 5
  },
  {
    weekStartDate: "2017-05-10",
    selectedDate: "2017-05-12",
    isoStartDay: 1
  },
  {
    weekStartDate: "2017-05-10",
    selectedDate: "2017-05-08",
    isoStartDay: 1
  },
  {
    weekStartDate: "2018-04-13",
    selectedDate: "2018-04-09",
    isoStartDay: 5
  },
  {
    weekStartDate: "1999-02-28",
    selectedDate: "1999-03-01",
    isoStartDay: 3
  },
  {
    weekStartDate: "2010-11-06",
    selectedDate: "2010-11-04",
    isoStartDay: 6
  },
  {
    weekStartDate: "2025-03-01",
    selectedDate: "2010-03-02",
    isoStartDay: 2
  },
  {
    weekStartDate: "2025-03-01",
    selectedDate: "2025-02-28",
    isoStartDay: 5
  }
];

describe("DateSlider _generateArrSelected", () => {
  const dateSlider = new DateSlider({});

  testData.forEach(item => {
    it(`should match snapshot when given a week startDate of ${item.weekStartDate} and a selectedDate of ${item.selectedDate} and the isoStartDay is ${item.isoStartDay}`, () => {
      const result = dateSlider._generateArrSelected(
        DateFns.getStartOfWeek(moment(item.weekStartDate), item.isoStartDay),
        DateFns.getStartOfWeek(moment(item.selectedDate), item.isoStartDay),
        moment(item.selectedDate),
        item.isoStartDay
      );
      expect(JSON.stringify(result)).toMatchSnapshot();
    });
  });
});
