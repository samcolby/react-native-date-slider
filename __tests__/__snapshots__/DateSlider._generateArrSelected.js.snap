// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 1999-02-28 and a selectedDate of 1999-03-01 and the isoStartDay is 3 1`] = `"[false,false,false,false,false,true,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 1999-02-28 and a selectedDate of 1999-03-12 and the isoStartDay is 3 1`] = `"[false,false,false,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2010-11-06 and a selectedDate of 2010-10-01 and the isoStartDay is 6 1`] = `"[false,false,false,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2010-11-06 and a selectedDate of 2010-11-04 and the isoStartDay is 6 1`] = `"[false,false,false,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2017-05-10 and a selectedDate of 2017-05-08 and the isoStartDay is 1 1`] = `"[true,false,false,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2017-05-10 and a selectedDate of 2017-05-10 and the isoStartDay is 1 1`] = `"[false,false,true,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2017-05-10 and a selectedDate of 2017-05-12 and the isoStartDay is 1 1`] = `"[false,false,false,false,true,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2017-05-10 and a selectedDate of 2017-05-21 and the isoStartDay is 1 1`] = `"[false,false,false,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2018-04-13 and a selectedDate of 2018-04-09 and the isoStartDay is 5 1`] = `"[false,false,false,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2018-04-13 and a selectedDate of 2018-04-14 and the isoStartDay is 5 1`] = `"[false,true,false,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2025-03-01 and a selectedDate of 2010-03-02 and the isoStartDay is 2 1`] = `"[false,false,false,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2025-03-01 and a selectedDate of 2010-10-01 and the isoStartDay is 2 1`] = `"[false,false,false,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2025-03-01 and a selectedDate of 2025-02-28 and the isoStartDay is 5 1`] = `"[true,false,false,false,false,false,false]"`;

exports[`DateSlider _generateArrSelected should match snapshot when given a week startDate of 2025-03-01 and a selectedDate of 2025-03-12 and the isoStartDay is 5 1`] = `"[false,false,false,false,false,false,false]"`;
